test/e2e/storage/testsuites/capacity.go

```diff
@@ -0,0 +1,380 @@
+	patterns := []storageframework.TestPattern{
+		storageframework.DefaultFsDynamicPV,
+	}
+	return InitCustomCapacityTestSuite(patterns)
+}
+
+func (p *capacityTestSuite) GetTestSuiteInfo() storageframework.TestSuiteInfo {
+	return p.tsInfo
+}
+
+func (p *capacityTestSuite) SkipUnsupportedTests(driver storageframework.TestDriver, pattern storageframework.TestPattern) {
+	// Check preconditions.
+	if pattern.VolType != storageframework.DynamicPV {
+		e2eskipper.Skipf("Suite %q does not support %v", p.tsInfo.Name, pattern.VolType)
+	}
+	dInfo := driver.GetDriverInfo()
+	if !dInfo.Capabilities[storageframework.CapCapacity] {
+		e2eskipper.Skipf("Driver %s doesn't publish storage capacity -- skipping", dInfo.Name)
+	}
+}
+
+func (p *capacityTestSuite) DefineTests(driver storageframework.TestDriver, pattern storageframework.TestPattern) {
+	var (
+		dInfo         = driver.GetDriverInfo()
+		dDriver       storageframework.DynamicPVTestDriver
+		driverCleanup func()
+		sc            *storagev1.StorageClass
+	)
+
+	// Beware that it also registers an AfterEach which renders f unusable. Any code using
+	// f must run inside an It or Context callback.
+	f := framework.NewFrameworkWithCustomTimeouts("capacity", storageframework.GetDriverTimeouts(driver))
+
+	init := func() {
+		dDriver, _ = driver.(storageframework.DynamicPVTestDriver)
+		// Now do the more expensive test initialization.
+		config, cleanup := driver.PrepareTest(f)
+		driverCleanup = cleanup
+		sc = dDriver.GetDynamicProvisionStorageClass(config, pattern.FsType)
+		if sc == nil {
+			e2eskipper.Skipf("Driver %q does not define Dynamic Provision StorageClass - skipping", dInfo.Name)
+		}
+	}
+
+	cleanup := func() {
+		err := storageutils.TryFunc(driverCleanup)
+		driverCleanup = nil
+		framework.ExpectNoError(err, "while cleaning up driver")
+	}
+
+	ginkgo.It("provides storage capacity information", func() {
+		init()
+		defer cleanup()
+
+		timeout := time.Minute
+		pollInterval := time.Second
+		matchSC := HaveCapacitiesForClass(sc.Name)
+		listAll := gomega.Eventually(func() (*storagev1beta1.CSIStorageCapacityList, error) {
+			return f.ClientSet.StorageV1beta1().CSIStorageCapacities("").List(context.Background(), metav1.ListOptions{})
+		}, timeout, pollInterval)
+
+		// If we have further information about what storage
+		// capacity information to expect from the driver,
+		// then we can make the check more specific. The baseline
+		// is that it provides some arbitrary capacity for the
+		// storage class.
+		matcher := matchSC
+		if dInfo.Capabilities[storageframework.CapCSILocalStorage] {
+			// Local storage. We expect one CSIStorageCapacity object per
+			// node for the storage class.
+			if len(dInfo.TopologyKeys) != 1 {
+				framework.Failf("need exactly one topology key for local storage, DriverInfo.TopologyKeys has: %v", dInfo.TopologyKeys)
+			}
+			matcher = HaveCapacitiesForClassAndNodes(f.ClientSet, sc.Provisioner, sc.Name, dInfo.TopologyKeys[0])
```

Comment:
Thinking about ways to make capacities matchers more generic instead of special casing local storage: Instead of having `CapCSILocalStorage`, what if we pass `len(dInfo.TopologyKeys)` into `HaveCapacitiesForClassAndNodes()`, and inside that function set TopologyKey to that one entry if `len()` is 1, otherwise ignore? We can then get rid of HaveCapacitiesForClass.

Reply:
Ah I realized this alone wouldn't work because the local storage matcher implementation assumes a single topology key implies distinct topology values for each node. Looking deeper.